### PR TITLE
Fix notes context handling issues

### PR DESCRIPTION
**Pull request by windingwind**
fix: #4024

### CODE REVIEW

File: chrome/content/zotero/elements/contextPane.js [hunk 1]
```diff
@@ -218,15 +215,7 @@
...
-			let tabNotesDeck = currentNoteContext.querySelector('.zotero-context-pane-tab-notes-deck');
-			let selectedIndex = Array.from(tabNotesDeck.children).findIndex(x => x.getAttribute('data-tab-id') == reader.tabID);
-			if (selectedIndex != -1) {
-				tabNotesDeck.setAttribute('selectedIndex', selectedIndex);
-				currentNoteContext.mode = "childNote";
-			}
-			else {
-				currentNoteContext._restoreViewType();
-			}
+			currentNoteContext.switchToTab(reader.tabID);
```
Comment (windingwind) [inline chrome/content/zotero/elements/contextPane.js:218]: 1. Clean up the `_handleTabSelect` and move some dirty code to `NotesContext#switchToTab`

File: chrome/content/zotero/elements/notesContext.js [hunk 2]
```diff
@@ -44,12 +44,21 @@
 					</vbox>
 				</vbox>
 				<vbox class="zotero-context-note-container context-note-standalone">
-					<vbox class="zotero-context-pane-editor-parent-line"></vbox>
+					<vbox class="zotero-context-pane-editor-parent-line">
```
Comment (windingwind) [inline chrome/content/zotero/elements/notesContext.js:47]: 2. Reuse note editor title bar

File: chrome/content/zotero/elements/notesContext.js [hunk 3]
```diff
@@ -193,6 +201,11 @@
 			});
 		}
 
+		_initNoteEditor() {
```
Comment (windingwind) [inline chrome/content/zotero/elements/notesContext.js:204]: 3. Only initialize note return button callback once

File: chrome/content/zotero/elements/notesContext.js [hunk 4]
```diff
@@ -247,7 +260,18 @@
 		_getCurrentEditor() {
 			let splitter = ZoteroContextPane.splitter;
 			if (splitter.getAttribute('state') == 'collapsed' || ZoteroContextPane.context.mode != "notes") return null;
-			return this.node.selectedPanel.querySelector('note-editor');
+			switch (this.mode) {
```
Comment (windingwind) [inline chrome/content/zotero/elements/notesContext.js:263]: 4. Fix `_getCurrentEditor` bug mentioned in #4036

File: chrome/content/zotero/elements/notesContext.js [hunk 5]
```diff
@@ -270,66 +292,82 @@
...
-			container.append(returnBtn, title);
-			parentTitleContainer.replaceChildren(container);
-			ZoteroContextPane.updateAddToNote();
 		}
 
 		updateNotesListFromCache() {
 			this._updateNotesList(true);
 		}
 
+
```
Comment (windingwind) [inline chrome/content/zotero/elements/notesContext.js:348]: 5. Make the logic of deciding what to show in the notes context (standalone note, current item note, or notes list) clearer and some bug fixes

**Review by windingwind — COMMENTED** [review]
The changes are marked in the comments.

**Comment by windingwind** [comment]
> Item note should be individual for each tab. Instead, it currently functions like a standalone note where the last opened note is visible in all tabs. It's best to check Zotero 6 to see how that should work.

What do you mean by `be individual`?

I need to figure out what to check in Z6. The current rules are as follows, I would appreciate it if you could give some hints on which part is inconsistent with our previous Z6 behavior @mrtcode :

1. If find an opened item note editor corresponds to the current tab, show the item note;
2. Otherwise, if find an opened standalone note editor, show the standalone note;
3. Otherwise, show the list.
4. When clicking the return button, remove the current note editor.

Say we have 3 tabs of reader: A, B, and C
1. if I open an item note $N_A$ in A and an item note $N_B$ in B, switching between these tabs I always get the $N_A$ in tab A, $N_B$ in tab B, and notes list in tab C;
2. if I open an item note $N_A$ in A and a standalone note $N_S$ in B, switching between these tabs I always get the $N_A$ in tab A and $N_S$ in tab B & C;

**Comment by mrtcode** [comment]
Opening an "Item Note" sets it only for the current tab.
Opening an "All Notes" note sets it for all tabs that have don't have "Item Note" set.
Pressing back from an "Item Note" shows notes list but if you don't open anything, the next time the tab is selected you will see an "All Notes" note or just the list if no "All Notes" note is set.
Pressing back from "All Notes" unsets "All Notes" note.

But again, we should use Zotero 6 as a reference here. This behavior was broken multiple times in Zotero 7 beta, therefore I think it would make sense to add a note at the top of context pane file explaining how it should work.

**Comment by dstillman** [comment]
(I actually saw some breakage in the latest version of Z6 as well when testing this.)

**Comment by windingwind** [comment]
> Opening an "Item Note" sets it only for the current tab.
Opening an "All Notes" note sets it for all tabs that have don't have "Item Note" set.
Pressing back from an "Item Note" shows notes list but if you don't open anything, the next time the tab is selected you will see an "All Notes" note or just the list if no "All Notes" note is set.
Pressing back from "All Notes" unsets "All Notes" note.

Yes, I think that's exactly what this PR is doing. Do you have steps to reproduce that breaks these rules?

**Comment by mrtcode** [comment]
> Yes, I think that's exactly what this PR is doing. Do you have steps to reproduce that breaks these rules?

Sorry, my fault. This PR actually fixes the mentioned issues. We can merge.

It was silently failing to build because of that sha256sum issue…

**Pull request merged**